pkg/cli/adbtunnel.go

```diff
@@ -252,38 +277,53 @@ type StatusMsg struct {
 }
 
 func (f *ADBForwarder) StatusJSON() []byte {
-	f.statusMtx.Lock()
-	status := f.status
-	f.statusMtx.Unlock()
+	// Pass equal values to get the current state without chaning it
+	_, status := f.compareAndSwapStatus(-1, -1)
 
 	msg := StatusMsg{
 		ServiceURL: f.serviceURL,
 		Zone:       f.zone,
 		Host:       f.host,
 		Device:     f.device,
 		Port:       f.port,
-		Status:     status,
+		Status:     StatusAsStr(status),
 	}
 	ret, err := json.Marshal(&msg)
 	if err != nil {
-		panic("Couldn't marshal status map")
+		panic(fmt.Sprintf("Couldn't marshal status map: %v", err))
 	}
 	return ret
 }
 
-func (f *ADBForwarder) setStatus(status string) {
+// Changes f.status to newStatus if it had oldStatus. Returns whether the change was
+// made and the old status.
+func (f *ADBForwarder) compareAndSwapStatus(oldStatus, newStatus int) (bool, int) {
```

Comment:
nit: use `current` instead of `old`, the method name could be: `swapStatusIfCurrent(current, new int)`

Reply:
compare_and_swap(old, new) is sort of a convention: https://en.wikipedia.org/wiki/Compare-and-swap

Reply:
I see, never mind then.